docs/reference/programming_model.rst

```diff
@@ -0,0 +1,259 @@
+.. meta::
+  :description: This chapter describes the HIP programming model, the contract
+                between the programmer and the compiler/runtime executing the
+                code.
+  :keywords: AMD, ROCm, HIP, CUDA, C++ language extensions
+
+*******************************************************************************
+Programming Model
+*******************************************************************************
+
+HIP defines a model of mapping SIMT programs (Single Instruction, Multiple
+Threads) onto various architectures, primarily GPUs. While the model may be
+expressed in most imperative languages, (eg. Python via PyHIP) this document
+will focus on the original C/C++ API of HIP.
+
+Threading Model
+===============================================================================
+
+The SIMT nature of HIP is captured by the ability to execute user-provided
+device programs, expressed as single-source C/C++ functions or sources compiled
+online/offline to binaries in bulk.
+
+Multiple instances of the device program (aka. kernel) may execute in parallel,
+all uniquely identified by a set of integral values which are referred to as
+thread IDs. The set of integers identifying a thread relate to the hierarchy in
+which threads execute.
+
+.. _inherent_thread_model:
+
+Inherent Thread Model
+-------------------------------------------------------------------------------
+
+The thread hiearchy inherent to how AMD GPUs operate manifest as depicted in
+:numref:`inherent_thread_hierarchy`.
+
+.. _inherent_thread_hierarchy:
+
+.. figure:: ../data/reference/programming_model/thread_hierarchy.svg
+  :alt: Diagram depicting nested rectangles of varying color. The outermost one
+        titled "Grid", inside sets of uniform rectangles layered on oneanother
+        titled "Block". Each "Block" containing sets of uniform rectangles
+        layered on oneanother titled "Warp". Each of the "Warp" titled
```

Comment:
Addendum: The passive voice might not be the best choice here, as it could also be misunderstood. Also the layering of the rectangles on top of each other might not be as important, the layout (4 rectangles layed out in 2 rows and 2 columns) also hasn't been explained, and I think that's not the important part this image has to convey, but rather the relation of blocks being contained in the grid and warps being contained in the block

Reply:
The layering wishes to accentuate the multi-dimensional indexing of blocks/threads.